### PR TITLE
feat!: Support URI encoded source maps

### LINKED ISSUE #59: sourceMappingURL with non-base64 data URI not supported.
```js
convert
  .fromComment('//# sourceMappingURL=data:application/json;charset=utf-8,%7B%22version%22:3,%22sources%22:%5B%22/Users/user/Development/node_modules/xyz/index.js%22%5D,%22names%22:%5B%5D,%22mappings%22:%22AAAA%22%7D')
  .toJSON()
```
Will throw:
```
SyntaxError: Unexpected token � in JSON at position 0
```
The map is:
```json
{"version":3,"sources":["/Users/user/Development/node_modules/xyz/index.js"],"names":[],"mappings":"AAAA"}
```
This effects https://github.com/Microsoft/vscode/issues/35978.

### PR DESCRIPTION
**Pull request by prantlf**
* Export a new regex compliant with RFC 2397.
* Recognise comments without base64 encoding.
* Add methods to convert from and to uri encoded strings.

Fixes #59 without dropping the compatibility.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -5,10 +5,24 @@ var SafeBuffer = require('safe-buffer');
...
+    return /^\s*\/(?:\/|\*)[@#]\s+sourceMappingURL=data:(((?:application|text)\/json)(?:;charset=[^;,]+?)?)?(?:;base64)?,.*$/mg;
+  }
+});
+
+Object.defineProperty(exports, 'commentRegex3', {
+  get: function getCommentRegex3 () {
+    // Groups: 1: media type, 2: MIME type, 3: charset, 4: encoding, 5: data.
+    return /^\s*\/(?:\/|\*)[@#]\s+sourceMappingURL=data:(((?:application|text)\/json)(?:;charset=([^;,]+)?)?)?(?:;(base64))?,(.*)$/;
+  }
+});
```
Comment (phated) [inline index.js:24]: @prantlf Why did you separate these two?

**Comment by thlorenz** [comment]
I like this in general, but we gotta be super careful here. Every time we change the regex it may break in certain scenarios.

If the RFC regex is more correct we should make it the default going forward and just provide the old one for people upgrading that run into issues.
However we have a PR open that reduces the chance of stackoverflows due to recursion when processing the regex (#65) and this RFC regex does not have those non-greedy features.

Did anyone actually complain ever that our existing regex doesn't match correctly?

**Comment by phated** [comment]
> Did anyone actually complain ever that our existing regex doesn't match correctly?

Jdalton complained in #59

**Comment by phated** [comment]
Since we are going to release this as a major. Can we remove the original `commentRegex` and combine the 2 new ones into a single? I still don't understand the need for two new RegExps

File: index.js (hunk 1 above)
Comment (phated) [reply index.js]: I combined them 😄

**Comment by phated** [comment]
@thlorenz I've rebased this and combined all the RegExp changes into a single modification of `commentRegex` that keeps the non-greedy changes from #65 - want to take a look?

Please note that this also removed the `:` separator for charsets because that isn't valid as per https://datatracker.ietf.org/doc/html/rfc2397 (but this was already going to be a breaking change). It cleaned up some of the test code.

**Comment by phated** [comment]
Here is a visualization of the new regexp
regexp

File: index.js [hunk 2]
```diff
@@ -66,8 +68,9 @@ function Converter (sm, opts) {
 
   if (opts.isFileComment) sm = readFromFileMap(sm, opts.commentFileDir);
   if (opts.hasComment) sm = stripComment(sm);
-  if (opts.isEncoded) sm = decodeBase64(sm);
-  if (opts.isJSON || opts.isEncoded) sm = JSON.parse(sm);
+  if (opts.encoding === 'base64') sm = decodeBase64(sm);
```
Comment (thlorenz) [inline index.js:71]: Let's add some braces here .. this is getting hard to parse

File: index.js [hunk 3]
```diff
@@ -104,9 +108,21 @@ function encodeBase64WithBtoa() {
...
+Converter.prototype.toURI = function () {
+  var json = this.toJSON();
+  return encodeURIComponent(json);
+};
+
 Converter.prototype.toComment = function (options) {
-  var base64 = this.toBase64();
-  var data = 'sourceMappingURL=data:application/json;charset=utf-8;base64,' + base64;
+  var encoding, content, data;
+  if (options && options.encoding === 'uri') {
```
Comment (thlorenz) [inline index.js:118]:
```suggestion
  if (options != null && options.encoding === 'uri') {
```
prefer to be explicit

File: index.js [hunk 4]
```diff
@@ -104,9 +108,21 @@ function encodeBase64WithBtoa() {
...
+    encoding = '';
+    content = this.toURI();
+  } else {
+    encoding = ';base64';
+    content = this.toBase64();
+  }
+  data = 'sourceMappingURL=data:application/json;charset=utf-8' + encoding + ',' + content;
   return options && options.multiline ? '/*# ' + data + ' */' : '//# ' + data;
```
Comment (thlorenz) [inline index.js:126]:
```suggestion
  return options != null && options.multiline ? '/*# ' + data + ' */' : '//# ' + data;
```
ditto

Comment (thlorenz) [reply index.js:126]: I know it was like this before but while we're at it ;)

File: test/comment-regex.js [hunk 5]
```diff
@@ -5,15 +5,27 @@ var test = require('tap').test
   , generator = require('inline-source-map')
   , convert = require('..')
 
-function comment(prefix, suffix) {
-  var rx = convert.commentRegex;
-  return rx.test(prefix + 'sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiIiwic291cmNlcyI6WyJmdW5jdGlvbiBmb28oKSB7XG4gY29uc29sZS5sb2coXCJoZWxsbyBJIGFtIGZvb1wiKTtcbiBjb25zb2xlLmxvZyhcIndobyBhcmUgeW91XCIpO1xufVxuXG5mb28oKTtcbiJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSJ9' + suffix)
+function comment(prefix, suffix, rx) {
+  return rx.exec(prefix + 'sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiIiwic291cmNlcyI6WyJmdW5jdGlvbiBmb28oKSB7XG4gY29uc29sZS5sb2coXCJoZWxsbyBJIGFtIGZvb1wiKTtcbiBjb25zb2xlLmxvZyhcIndobyBhcmUgeW91XCIpO1xufVxuXG5mb28oKTtcbiJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSJ9' + suffix)
```
Comment (thlorenz) [inline test/comment-regex.js:9]: I'm a bit worried about changing tests along with functionality since now we cannot be sure if the original behavior changed in an unintended way.
Is there a reason why this is needed?

File: test/comment-regex.js [hunk 6]
```diff
@@ -28,15 +40,21 @@ test('comment regex old spec - @', function (t) {
     '\t/*@ ', // multi line style with leading tab
     '/*@ ',   // multi line style with leading text
   ].forEach(function (x) { 
-    t.ok(comment(x, ''), 'matches ' + x) 
-    t.ok(commentWithCharSet(x, ''), 'matches ' + x + ' with charset') 
-    t.ok(commentWithCharSet(x, '', '='), 'matches ' + x + ' with charset')
+    t.ok(comment(x, '', convert.commentRegex), 'matches ' + x)
```
Comment (thlorenz) [inline test/comment-regex.js:43]: ditto here .. existing tests should be touched as little as possible to ensure nothing breaks

File: test/comment-regex.js [hunk 7]
```diff
@@ -51,15 +69,93 @@ test('comment regex new spec - #', function (t) {
     '\t/*# ', // multi line style with leading tab
     '/*# ',   // multi line style with leading text
   ].forEach(function (x) { 
-    t.ok(comment(x, ''), 'matches ' + x) 
-    t.ok(commentWithCharSet(x, ''), 'matches ' + x + ' with charset') 
-    t.ok(commentWithCharSet(x, '', '='), 'matches ' + x + ' with charset')
+    t.ok(comment(x, '', convert.commentRegex), 'matches ' + x)
```
Comment (thlorenz) [inline test/comment-regex.js:72]: ditto here a bunch of stuff changed .. maybe I'm just missing something and this is smaller than appears, but I still don't understand why this was needed

File: test/comment-regex.js [hunk 8]
```diff
@@ -51,15 +69,93 @@ test('comment regex new spec - #', function (t) {
...
+    '\t//# ', // with leading tab
+    '//# ',   // with leading text
+    '/*# ',   // multi line style
+    '  /*# ', // multi line style with leading spaces
+    '\t/*# ', // multi line style with leading tab
+    '/*# ',   // multi line style with leading text
+  ].forEach(function (x) {
+    var m;
+    m = comment(x, '', convert.commentRegex)
+    t.ok(m, 'matches ' + x)
```
Comment (thlorenz) [inline test/comment-regex.js:103]: It seems here we're accessing the matches so `rx.test` would not work, however we could have just added a custom method for this case and left the original ones alone

File: test/convert-source-map.js [hunk 9]
```diff
@@ -4,30 +4,45 @@
 var test = require('tap').test
   , generator = require('inline-source-map')
   , convert = require('..')
+  , decodeBase64 = typeof Buffer.from ?
+    function decodeBase64(base64) {
```
Comment (thlorenz) [inline test/convert-source-map.js:8]: This is super hard to parse. Please declare these methods separately and refer to them by name .. most likely this becomes a one-liner then

**Review by thlorenz — CHANGES_REQUESTED** [review]
Thanks for all the work putting this together.
Let's clean up the questions/requests I had and I'll have another look.

From what I can tell the regex looks fine, but just want to make sure the original behavior didn't change and we need the tests to be as unmodified as possible to be sure of that.

File: test/comment-regex.js (hunk 7 above)
Comment (phated) [reply test/comment-regex.js:72]: Oh yeah, this changed because different regexps were being passed to test them all, this can be reduced back to the original because it is one regexp again

File: test/comment-regex.js (hunk 8 above)
Comment (phated) [reply test/comment-regex.js:140]: I created a separate set of functions that use `exec` that only exist inside this test.

File: test/comment-regex.js [hunk 10]
```diff
@@ -10,10 +10,29 @@ function comment(prefix, suffix) {
   return rx.test(prefix + 'sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiIiwic291cmNlcyI6WyJmdW5jdGlvbiBmb28oKSB7XG4gY29uc29sZS5sb2coXCJoZWxsbyBJIGFtIGZvb1wiKTtcbiBjb25zb2xlLmxvZyhcIndobyBhcmUgeW91XCIpO1xufVxuXG5mb28oKTtcbiJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSJ9' + suffix)
 }
 
-function commentWithCharSet(prefix, suffix, sep) {
-  sep = sep || ':';
```
Comment (phated) [inline test/comment-regex.js:14]: The `:` separator was removed because that is not a valid separator between `charset` and `utf-8`. This also resulted in some tests being removed below.

File: index.js (hunk 2 above)
Comment (phated) [reply index.js]: Done!

File: test/comment-regex.js (hunk 5 above)
Comment (phated) [reply test/comment-regex.js]: Left the original alone and added new ones for the newly supported formats (that all use `.test`)

File: test/comment-regex.js (hunk 7 above)
Comment (phated) [reply test/comment-regex.js]: Reduced the diff!

File: test/convert-source-map.js (hunk 9 above)
Comment (phated) [reply test/convert-source-map.js]: Done!

**Comment by phated** [comment]
This introduces a subtle (but necessary) breaking change with the regexp. It will no longer parse a sourcemap that had the encoding specified at `charset:utf-8`; however, it doesn't seem that was ever (??) valid syntax for a data-uri, so I don't know if it will actually affect any users.

**Review by thlorenz — APPROVED** [review]
LGTM thanks for your patience and making this perfect :) 

Feel free to squash and merge + LMK your npm user and I'll add you so you can publish.

As discussed this needs to be a major bump since it breaks backwards compat.
